# Agenda

# 1.0. Problema de Negócio

https://sejaumdatascientist.com/os-5-projetos-de-data-science-que-fara-o-recrutador-olhar-para-voce/

**Empresa:** House Rocket  

**O que a empresa faz:** House Rocket é uma plataforma de compra e vendas de imóvies.  

**Qual o problema:** O CEO da empresa gostaria de maximizar os lucros da empresa encontrando bons negócios.  

**A principal estratégia:** O CEO possuí fontes externas de informação. Porém ele acredita que essas fontes não sejam a melhor opção. Dessa forma, ele deseja que seja verificado o Dataset local.  

**As perguntas do CEO da House Rocket:**
1. Quantas casas estão disponíveis para compra?
2. Quantos atributos as casas possuem? (número de quartos, tamanho (m2), etc)
3. Quais são esses atributos? (nome)
4. Qual a casa mais cara do portifólio?
5. Qual a casa com o maior número de quartos?

# 2.0. Planejamento de Solução

## 2.1. Planejamento do Produto Final

1. O que será entregue? (Planilha, Texto, Email, Modelos de ML, ... )  
    R: Texto com as Respostas


2. Com vai ser a Entregua?  
    R: Pergunta e Resposta
       
   Exemplo:  
    Quantas casas estão disponíveis para compra?  
    R: XXX Casas.

## 2.2. Planejamento do Processo

Para sabermos como será feito o processo, primeiro temos que responder algumas perguntas:  

**1. Onde Está a Informação? (Excel, BD, API, ...)**  
R: Os dados estão dentro de um arquivo .csv (https://www.kaggle.com/harlfoxem/housesalesprediction?select=kc_house_data.csv)

**2. Como coletar essas informações? (SQL, Python, Streamlit, ...)**  
R: Será feito o Download dos dados

**3. Começar a responder as perguntas**  
1. Quantas casas estão disponíveis para compra?  
R: É a mesma quantidade de linhas do dataset


2. Quantos atributos as casas possuem? (número de quartos, tamanho (m2), etc)  
R: É a mesma quantidade de colunas do dataset


3. Quais são esses atributos? (nome)  
R: É o nome das colunas do dataset


4. Qual a casa mais cara do portifólio?  
R: Será a primeira linha do dataset se ordenarmos o dataset pela coluna de valor em ordem descrescente


5. Qual a casa com o maior número de quartos?  
R: Será a primeira linha do dataset se ordenarmos o dataset pela coluna de quantidade de quartos em ordem descrescente

## 2.3. Planejamento das Ferramentas

Essa parte irá depender muito da Stack da empresa.  
No geral, é sempre utilizado uma linguagem de programação para resolver e criar soluções conforme as demandas da emrpesa.  
Porém, podem ser utilizados também ferramentas de *Office* como o Excel.

Vanteganes de se utilizar linguagens de programação como ferramenta:
* Foi criada para criar softwares
* Mais fácil de processar e manipular dados para criar *Insights*

# 3.0. O que é o Python?

Python é uma linguagem de programação desenvolvida por Guido van Rossum no final da década de 1980 com o objetivo de ser fácil e intuitiva, porém poderosa.

Suas principais características são:
* Linguagem de programação de alto nível
* Interpretada e de código-fonte aberto
* Interativa
* Multi-plataforma e Multi-paradigma
* Sintaxe simples, fácil de aprender e de manter
* Tipagem forte e dinâmica
* Tudo em Python é um objeto: variáveis, funções, etc. Cada objeto tem um ID, tipo e valor


Curiosamente o nome não tem nenhuma relação com o anfíbio de mesmo nome e sim uma homenagem ao grupo de comédia britânico Monty Python!

* Ambientes de Desenvolvimento para Python
    * PyCharm
    * IDLE
    * Anaconda


* IDE
    * Visual Studio Code
    * Atom


* Ambientes Interativos
    * IPython
    * Jupyter Notebook

Sistema de instalação e gerenciamento de pacotes do Python utiliza o Python Package Index ou PyPI como repositório
* https://pypi.org/


A grande vantagem da utilização do pip é a facilidade de sua execução através de linha de comando:
* pip install algum-nome-de-pacote
* pip uninstall algum-nome-de-pacote

# 4.0. Como escrever Código Python?

Incialmente, precisamos somente de duas coisas:
1. Editor de Texto
2. Interpretador do Python

Iremos utilizar o *Jupiter Notebook* como editor de texto e o Python3 como Interpretador

# 5.0. Os primeiros comandos Python

* A identação é usada em Python para delimitar blocos de código. O número de espaços pode variar, porém todos os statements (instruções) dentro de um mesmo bloco, devem possuir a mesma quantidade de espaços
* A primeira linha que compõe statements como *if*, *else*, *while*, *def*, e *class* deve ser finalizada com dois pontos (:)
* O ponto e vírgula é opcional ao final de um statement
* Comentários de apenas uma linha começam por sustenido (#); Comentários de multiplas linhas devem ter um sustenido (#) em cada linha 
* Arquivos de programas Python tem extensão .py

In [6]:
# Comentário de Linha

# Comentário
# de
# Bloco

In [5]:
# Exibe valores
print("Hello World! 123")

Hello World! 123


In [11]:
# Declaração de Variáveis
i = 10
i

10

In [15]:
# Verificar qual o tipo da variável
type(i)

int

## 5.1. Tipos de Dados

* **Numéricos:** Inteiros (Integer), Decimais (Floats), Frações (Fractions) e Números Complexos (Complex Numbers)  
* **Strings:** Sequencia de Caracteres  
* **Listas:** Sequência ordenada (possuem índice) de valores
* **Tuplas:** Sequência ordenada (possuem índice) imutável de valores
* **Dicionários:** Conjunto não ordenado de conteúdos do tipo chave-valor

### 5.1.1. Tipos Numéricos

Números são objetos imutáveis em Python, isso significa que seus valores não podem ser alterados  

Existem 3 tipos de dados para números em Python
* int (signed integer) -> Também chamado apenas de integer, comporta números positivos e negativos sem casas decimais
* float (floating point) -> Representa números reais e são escritos com ponto decimal, dividindo um inteiro em partes fracionárias, também podendo ser escrito em notação científica com “e” indicando potencia de 10 (ex:2.5e2)
* complex São escritos por dois valores reais, a parte real e a parte imaginária na forma (real + IMAG J). Neste caso o número i (√-1 ) é designado pela letra j. (Não são muito utilizados)

**Principais Funções para tipos numéricos**

Função | Descrição
:------------|:----------------
int (x) | Converte x em um inteiro
float (x) | Converte x em um ponto-flutuante
abs (x) | Retorna o valor absoluto de x
exp (x) | Retorna o exponencial de x (e^x)
log (x) | Retorna o logaritmo natural de x (inverso da função exponencial)
pow (x, y) | Retorna o valor de x elevado à potencia y
sqrt (x) | Retorna a raiz quadrada de x
round (x, y) | Retorna x arredondado em y casas decimais

### 5.1.2. Tipo String

* String, assim como numbers, são objetos imutáveis em Python. Dessa forma um update só pode ser possível com a alocação de um novo objeto com o novo conteúdo
* Python não suporta tipos “char”. Um char em Python é considerado uma String de tamanho 1 
* Uma String pode ser atribuída de várias formas diferentes
    * ‘Uma String’
    * “Outro exemplo de String”
    * ’’’String com multiplas linhas’’’
* Strings iniciam sempre com indice 0
* Pode se usar operações como slicing ([], [:]), concatenação (+), repetição (*) e membership (in)

**Funções e métodos mais comuns com tipo String**


Função | Descrição
:--------|:-----------
str (num) | Converte um número em String
len (str) | Retorna o tamanho de uma String
str.count (s) | Retorna a quatidade de conjuntos s presentes na string
str.isalpha () | Retorna False se a string contiver algum caracter que não seja letras
str.isdigit () | Retorna False se a string contiver algum caracter que não seja número
str.lower () | Retorna a string transformada em minúsculos
str.upper () | Retorna a string transformada em maiúsculos
str.replace (old, new) | Substitui uma porção da string por outro conteúdo
str.strip () | Retira espaços em branco no começo e no fim da string
str.title () | Retorna a string capitalizada (iniciais em maiúscula)
str.split (delimit) | Separa uma string conforme um delimitador. É o inverso do join()
str.join (sequence) | Junta cada item da string com um delimitador especificado. É o inverso do split().

### 5.1.3.Listas

* Listas são grupos de itens ou elementos indexados, não necessariamente do mesmo tipo
* Listas são mutáveis, isso significa que seu conteúdo pode ser alterado sem que um novo objeto seja criado
* Para atribuir elementos a uma lista é necessário que estejam entre colchetes e separados por vírgula, ex:
    * l = [item1, item2, ... Itemn]
* Os indices em uma lista sempre se iniciam em 0 (zero)
* Listas podem conter sublistas que podem conter sublistas e assim por diante, tornando esse tipo de dado extremamente versátil
* Pode se usar operações como slicing ([], [:]), concatenação (+), repetição (*) e membership (in)

**Funções e métodos mais comuns com tipo Lista**

Função | Descrição
:--------|:----------
len(lista) | Retorna o tamanho total da lista
max(lista) | Retorna o maior elemento da lista
min(lista) | Retorna o menor elemento da lista
list(tupla) | Converte uma tupla em uma lista
list.append(obj) | Adiciona um objeto à lista
list.insert(index, obj) | Insere um objeto à lista em determinada posição
list.count(obj) | Retorna a quantidade de vezes que um determinado objeto ocorre na lista
list.index(obj) |Retorna o primeiro indice de ocorrencia de um determinado objeto
list.remove(obj) | Remove um objeto da lista
list.reverse() | Reverta o ordenamento da lista
list.sort() | Ordena a lista

### 5.1.4. Tuplas

* Tuplas são tipos muito semelhantes às listas, porém imutáveis, por esse motivo, tuplas são mais eficientes que as listas. Dessa forma quando não existe a necessidade de alteração de seu conteúdo, devem ser escolhidas preferencialmente
* Como são imutávais, protegem os dados contidos nela de alterações acidentais
* Para atribuir elementos a uma tupla é necessário que estejam entre parêntesis e separados por vírgula, ex:
    * t = (item1, item2, ... Itemn)
* Para tuplas de um elemento apenas é necessário incluir a vírgula após o elemento:
    * t = (item1,)
* Indices funcionam da mesma forma das listas, assim como os métodos e funções

### 5.1.5. Dicionários

* Dicionários são uma implementação de hash table que consiste em uma lista de pares de chave-valor, sem ordenação
* Chaves, devem ser tipos imutáveis e usualmente são utilizadas strings ou números
* Valores podem ser qualquer tipo de objeto Python
* Para atribuir elementos a um dicionário é necessário que estejam entre chaves ( { } ), separados por vírgula ( , ) , e com os pares chave, valor separados por dois pontos (:)
    * dic = {‘k1’:1, ‘k2’:'Pedro', ... ‘kn’:n}
* Os elementos de um dicionário podem ser acessados ou alterados através de sua chave entre colchetes
    * dic[‘k1’] = 2

**Funções métodos mais comuns com tipo Dicionário**

Função | Descrição
:----------|:---------
len(dict) | Retorna o número de elementos do dicionário
str(dict) | Retorna a representação em formato string do dicionário
dict.keys() | Retorna a lista de chaves do dicionário
dict.values() | Retorna a lista de valores do dicionário
dict.items() | Retorna a lista de chave, valor do dicionário
dict.get(key, default=None) | Retorna o valor de uma chave ou um valor default caso não seja encontrada
dict.update(dict2) | Insere um elemento (chave, valor) no dicionário
dict.clear() | Remove todos os elementos do diciónário

### 5.2. Importando Pacotes/Bibliotecas

Muitas funções do Python estão disponíveis em bibliotecas ou pacotes. Para utilizarmos essas funções, basta importarmos um determinado pacote:

**import pacote as apelido**

Caso esse pacote não esteja instalado, basta instalá-lo utilizando o gerenciador de pacotes do Python, o pip:

**pip install pacote**

## 6.0. Responder as Questões de Negócio

Para iniciarmos a responder as questões, vamos inicialmente ler a base de dados. Para isso, precisaremos da biblioteca *pandas*, pois ela possuí várias funções que nos auxiliarão tanto na leitura quanto na manipulação e tratamento dos dados

In [45]:
# Instalando o Pandas na máquina
!pip install pandas

In [46]:
# Importando a biblioteca para que ela possa ser utilizada
import pandas as pd

Com os a biblioteca importada, vamos agora ler o conjunto de dados do arquivo .csv:

In [47]:
# Lendo os dados do dataset 
df = pd.read_csv('./datasets/kc_house_data.csv')

In [53]:
# Exibe todo o dataset
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

Vamos executar os comandos necessários para iniciar as respostas das perguntas do CEO da House Rocket:

In [49]:
# Mostra as 5 primeiras linhas do dataset
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [51]:
# Mostra o número de linhas e colunas do dataset
df.shape

(21613, 21)

In [54]:
# Monstra os títulos das colunas do dataset
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [58]:
# Ordena de forma decrescente o dataset pela coluna price e exibe somente as colunas ID e Price
print(df[['id', 'price']].sort_values('price', ascending=False))

               id      price
7252   6762700020  7700000.0
3914   9808700762  7062500.0
9254   9208900037  6885000.0
4411   2470100110  5570000.0
1448   8907500070  5350000.0
...           ...        ...
8274   3883800011    82000.0
16198  3028200080    81000.0
465    8658300340    80000.0
15293    40000362    78000.0
1149   3421079032    75000.0

[21613 rows x 2 columns]


In [59]:
# Ordena de forma decrescente o dataset pela coluna bedrooms e exibe somente as colunas ID e Bedrooms
print(df[['id', 'bedrooms']].sort_values('bedrooms', ascending=False))

               id  bedrooms
15870  2402100895        33
8757   1773100755        11
15161  5566100170        10
13314   627300145        10
19254  8812401450        10
...           ...       ...
19452  3980300371         0
8484   2310060040         0
875    6306400140         0
8477   2569500210         0
9773   3374500520         0

[21613 rows x 2 columns]


Com as informações em mãos, basta responder as perguntas:

1. Quantas casas estão disponíveis para compra?  
R: Existem 21.613 casas disponíveis

2. Quantos atributos as casas possuem? (número de quartos, tamanho (m2), etc)  
R: As casas possuem 21 atributos

3. Quais são esses atributos? (nome)  
R: Os nomes dos atributos são: 'id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', 'sqft_lot15'

4. Qual a casa mais cara do portifólio?  
R: A Casa com o maior valor é a casa que possuí o ID 6762700020 que custa R$ 7.700.000,00

5. Qual a casa com o maior número de quartos?  
R: A casa com o maior números de quartos é a casa que possuí o ID 2402100895 que tem 33 quartos

## 7.0. Exercícios

#### 6. Qual a soma total de quartos do conjunto de dados?

In [85]:
bedrooms = df[['bedrooms']].sum()
bedrooms

bedrooms    72854
dtype: int64

Qual a soma total de quartos do conjunto de dados?  
R: 72.854 quartos

#### 7. Quantas casas possuem 2 banheiros?

In [94]:
bathrooms = df.loc[df['bathrooms'] == 2.00]
bathrooms.count()

id               1930
date             1930
price            1930
bedrooms         1930
bathrooms        1930
sqft_living      1930
sqft_lot         1930
floors           1930
waterfront       1930
view             1930
condition        1930
grade            1930
sqft_above       1930
sqft_basement    1930
yr_built         1930
yr_renovated     1930
zipcode          1930
lat              1930
long             1930
sqft_living15    1930
sqft_lot15       1930
dtype: int64

Quantas casas possuem 2 banheiros?  
R: 1930 Casas

#### 8. Qual o preço médio de todas as casas no conjunto de dados?

In [84]:
mean_price = df[['price']].mean()
round(mean_price, 2)

price    540088.14
dtype: float64

Qual o preço médio de todas as casas no conjunto de dados?  
R: O Valor médio das casas é de R$ 540.088,14

#### 9. Qual o preço médio de casas com 2 banheiros?

In [98]:
bathrooms.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
4   1954400510  20150218T000000  510000.0         3        2.0         1680   
14  1175000570  20150312T000000  530000.0         5        2.0         1810   
16  1875500060  20140731T000000  395000.0         3        2.0         1890   
25  1202000200  20141103T000000  233000.0         3        2.0         1710   
37  2768000400  20141230T000000  640000.0         4        2.0         2360   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
4       8080     1.0           0     0  ...      8        1680              0   
14      4850     1.5           0     0  ...      7        1810              0   
16     14040     2.0           0     0  ...      7        1890              0   
25      4697     1.5           0     0  ...      6        1710              0   
37      6000     2.0           0     0  ...      8        2360              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
4       1987             0    98074  47.6168 -122.045           1800   
14      1900             0    98107  47.6700 -122.394           1360   
16      1994             0    98019  47.7277 -121.962           1890   
25      1941             0    98002  47.3048 -122.218           1030   
37      1904             0    98107  47.6702 -122.362           1730   

    sqft_lot15  
4         7503  
14        4850  
16       14018  
25        4705  
37        4700  

[5 rows x 21 columns]

In [97]:
round(bathrooms[['price']].mean(), 2)

price    457889.72
dtype: float64

Qual o preço médio de casas com 2 banheiros?  
R: O preço médio é de R$ 457.889,72

#### 10. Qual o preço mínimo entre as casas com 3 quartos?

In [100]:
bedrooms_value = df.loc[df['bedrooms'] == 3]
bedrooms_value.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   
6  1321400060  20140627T000000  257500.0         3       2.25         1715   
7  2008000270  20150115T000000  291850.0         3       1.50         1060   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
4      8080     1.0           0     0  ...      8        1680              0   
6      6819     2.0           0     0  ...      7        1715              0   
7      9711     1.0           0     0  ...      7        1060              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
4      1987             0    98074  47.6168 -122.045           1800   
6      1995             0    98003  47.3097 -122.327           2238   
7      1963             0    98198  47.4095 -122.315           1650   

   sqft_lot15  
0        5650  
1        7639  
4        7503  
6        6819  
7        9711  

[5 rows x 21 columns]

In [101]:
bedrooms_value[['price']].min()

price    82000.0
dtype: float64

Qual o preço mínimo entre as casas com 3 quartos?  
R: O preço mínimo é de R$ 82.000,00

#### 11. Quantas casas possuem mais de 300 metros quadrados na sala de estar?

In [104]:
sqrt_living = df.loc[df['sqft_living'] >= 300]
sqrt_living[['id', 'sqft_living', 'price']]

id  sqft_living     price
0      7129300520         1180  221900.0
1      6414100192         2570  538000.0
2      5631500400          770  180000.0
3      2487200875         1960  604000.0
4      1954400510         1680  510000.0
...           ...          ...       ...
21608   263000018         1530  360000.0
21609  6600060120         2310  400000.0
21610  1523300141         1020  402101.0
21611   291310100         1600  400000.0
21612  1523300157         1020  325000.0

[21612 rows x 3 columns]

Quantas casas possuem mais de 300 metros quadrados na sala de estar?  
R: 21.612 casas

#### 12. Quantas casas tem mais de 2 andares?

In [113]:
floors = df.loc[df['floors'] >= 2]
floors[['id', 'floors', 'price']]

id  floors     price
1      6414100192     2.0  538000.0
6      1321400060     2.0  257500.0
9      3793500160     2.0  323000.0
15     9297300055     2.0  650000.0
16     1875500060     2.0  395000.0
...           ...     ...       ...
21608   263000018     3.0  360000.0
21609  6600060120     2.0  400000.0
21610  1523300141     2.0  402101.0
21611   291310100     2.0  400000.0
21612  1523300157     2.0  325000.0

[9023 rows x 3 columns]

Quantas casas tem mais de 2 andares?  
R: 9.023 casas

#### 13. Quantas casas tem vista para o mar?

In [120]:
waterfront = df.loc[df['waterfront'] != 0]
waterfront[['id', 'price', 'waterfront']]

id      price  waterfront
49      822039084  1350000.0           1
230    8096000060   655000.0           1
246    2025069065  2400000.0           1
264    2123039032   369900.0           1
300    3225069065  3075000.0           1
...           ...        ...         ...
19984  2025069140  1898000.0           1
20325   518500480  3000000.0           1
20767  8043700105  2300000.0           1
21201   518500460  2230000.0           1
21576  9253900271  3567000.0           1

[163 rows x 3 columns]

Quantas casas tem vista para o mar?  
R: 163 casas

#### 14. Das casas com vista para o mar, quantas tem 3 quartos?

In [119]:
waterfront_bedrooms = waterfront.loc[df['bedrooms'] == 3]
waterfront_bedrooms[['id', 'price', 'bedrooms']]

id      price  bedrooms
49      822039084  1350000.0         3
457    2122039094   705000.0         3
540     622049114  2125000.0         3
656    3760500116  3070000.0         3
1261   7636800041   995000.0         3
...           ...        ...       ...
19622  7351200295  1150000.0         3
19971  2623039019   988500.0         3
19984  2025069140  1898000.0         3
20325   518500480  3000000.0         3
21201   518500460  2230000.0         3

[64 rows x 3 columns]

Das casas com vista para o mar, quantas tem 3 quartos?  
R: 64 casas

#### 15. Das casas com mais de 300 metros quadrados de sala de estar, quantas tem mais de 2 banheiros?

In [122]:
sqrt_living_bathrooms = sqrt_living.loc[sqrt_living['bathrooms'] >= 2]
sqrt_living_bathrooms[['id', 'price', 'sqft_living', 'bathrooms']]

id      price  sqft_living  bathrooms
1      6414100192   538000.0         2570       2.25
3      2487200875   604000.0         1960       3.00
4      1954400510   510000.0         1680       2.00
5      7237550310  1225000.0         5420       4.50
6      1321400060   257500.0         1715       2.25
...           ...        ...          ...        ...
21606  7936000429  1007500.0         3510       3.50
21607  2997800021   475000.0         1310       2.50
21608   263000018   360000.0         1530       2.50
21609  6600060120   400000.0         2310       2.50
21611   291310100   400000.0         1600       2.50

[13172 rows x 4 columns]

Das casas com mais de 300 metros quadrados de sala de estar, quantas tem mais de 2 banheiros?  
R: 13.172 casas

# Correção do Meigarom

In [175]:
!pip install pandas numpy

In [176]:
import pandas as pd
import numpy as np

In [126]:
data = pd.read_csv('./datasets/kc_house_data.csv')

In [127]:
data.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# 1. Quantas casas estão disponíveis para compra?

In [129]:
# Estratégia:
# 1. Selecionar a coluna ID
# 2. Contar o número de valores únicos

houses = len( data['id'].unique() )
houses

21436

In [131]:
len( data['id'].drop_duplicates() )

21436

In [132]:
print( 'O número de casas disponíveis é de {}'.format( houses ) )

O número de casas disponíveis é de 21436


# 2. Quantos atributos as casas possuem? (número de quartos, tamanho (m2), etc)

In [138]:
# O numero de colunas representam os atributos do apartamento.
# id e date - não são atributos do apartamento

quantity_attributes = len( data.drop( ['id', 'date'], axis=1 ).columns )
quantity_attributes

19

In [139]:
print( 'A quantidade de atributos das casas é de {}'.format( quantity_attributes ) )

A quantidade de atributos das casas é de 19


# 3. Quais são esses atributos? (nome)

In [156]:
attributes = data.drop( ['id', 'date'], axis=1 )
attributes.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0         3       1.00         1180      5650     1.0           0   
1  538000.0         3       2.25         2570      7242     2.0           0   
2  180000.0         2       1.00          770     10000     1.0           0   
3  604000.0         4       3.00         1960      5000     1.0           0   
4  510000.0         3       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode      lat     long  sqft_living15  sqft_lot15  
0    98178  47.5112 -122.257           1340        5650  
1    98125  47.7210 -122.319           1690        7639  
2    98028  47.7379 -122.233           2720        8062  
3    98136  47.5208 -122.393           1360        5000  
4    98074  47.6168 -122.045           1800        7503

In [155]:
print( 'Os atributos das casas são: {}'.format( attributes.columns.tolist() ))

Os atributos das casas são: ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', 'sqft_lot15']


# 4. Qual a casa mais cara do portifólio?

In [157]:
expensive = data[['id', 'price']].sort_values( 'price', ascending=False ).reset_index( drop=True ).loc[0, ['id', 'price']]
expensive

id       6.762700e+09
price    7.700000e+06
Name: 0, dtype: float64

In [158]:
print( 'A casa mais cara do portifólio é a casa de ID {} com o valor de R$ {}'.format( expensive['id'], round(expensive['price'], 2) ))

A casa mais cara do portifólio é a casa de ID 6762700020.0 com o valor de R$ 7700000.0


# 5. Qual a casa com o maior número de quartos?

In [162]:
bedrooms = data[['id', 'bedrooms']].sort_values( 'bedrooms', ascending=False )\
                                    .reset_index( drop=True ).loc[0, ['id', 'bedrooms']]
bedrooms

id          2402100895
bedrooms            33
Name: 0, dtype: int64

In [164]:
print( 'A casa com o maior número de quartos é a casa de ID {} com {} quartos'.format( bedrooms['id'], round(bedrooms['bedrooms'], 2) ))

A casa com o maior número de quartos é a casa de ID 2402100895 com 33 quartos


# 6. Qual a soma total de quartos do conjunto de dados?

In [165]:
num_bedrooms = data['bedrooms'].sum()
num_bedrooms

72854

In [167]:
print( 'A soma total de quartos na base de dados é de {}'.format( num_bedrooms ))

A soma total de quartos na base de dados é de 72854


# 7. Quantas casas possuem 2 banheiros?

In [173]:
# Estratégia
# 1. Filtrar somente os id que possuem valor da coluna bathroom igual a 2
# 2. Selecionar as colunas id e bathroom
# 2. Contar o número de valores da coluna id 

num_bathrooms = len (data.loc[data['bathrooms'] == 2, ['id', 'bathrooms']] )
num_bathrooms

1930

In [174]:
print( 'O número de casas que possuem 2 banheiros é {}'.format( num_bathrooms ))

O número de casas que possuem 2 banheiros é 1930


# 8. Qual o preço médio de todas as casas no conjunto de dados?

In [178]:
mean_price = np.round( data['price'].mean(), 2 )
mean_price

540088.14

In [180]:
print( 'O preço médio das casa é R$ {}'.format( mean_price ))

O preço médio das casa é R$ 540088.14


# 9. Qual o preço médio de casas com 2 banheiros?

In [182]:
mean_price_2 = np.round( data.loc[ data['bathrooms'] == 2, 'price'].mean(), 2 )
mean_price_2

457889.72

In [183]:
print( 'O preço médio das casa com 2 banheiros é R$ {}'.format( mean_price_2 ))

O preço médio das casa com 2 banheiros é R$ 457889.72


# 10. Qual o preço mínimo entre as casas com 3 quartos?

In [189]:
min_price = data.loc[ data['bedrooms'] == 3, 'price'].min()
min_price

82000.0

In [190]:
print( 'O preço mínimo das casa com 3 quartos é R$ {}'.format( min_price ))

O preço mínimo das casa com 3 quartos é R$ 82000.0


# 11. Quantas casas possuem mais de 300 metros quadrados na sala de estar?

In [187]:
data['m2'] = data['sqft_living'] * 0.093

living = len( data.loc[ data['m2'] > 300, 'id'] ) 
living

2258

In [188]:
print( 'A quantidade de casas que possuem mais de 300 metros quadrados de sala de estar é de {}'.format( living ))

A quantidade de casas que possuem mais de 300 metros quadrados de sala de estar é de 2258


# 12. Quantas casas tem mais de 2 andares?

In [193]:
floors = len( data.loc[ data['floors'] > 2, 'id'] ) 
floors

782

In [194]:
print( 'A quantidade de casas que possuem mais de 2 andares é de {}'.format( floors ))

A quantidade de casas que possuem mais de 2 andares é de 782


# 13. Quantas casas tem vista para o mar?

In [195]:
water = len( data.loc[ data['waterfront'] == 1, 'id'] ) 
water

163

In [196]:
print( 'A quantidade de casas que possuem vista para o mar é de {}'.format( water ))

A quantidade de casas que possuem vista para o mar é de 163


# 14. Das casas com vista para o mar, quantas tem 3 quartos?

In [198]:
water_2 = data.loc[((data['waterfront'] == 1) & (data['bedrooms'] == 3)), 'id'].size
water_2

64

In [199]:
print( 'A quantidade de casas que possuem vista para o mar e tem 3 quartos é de {}'.format( water_2 ))

A quantidade de casas que possuem vista para o mar e tem 3 quartos é de 64


# 15. Das casas com mais de 300 metros quadrados de sala de estar, quantas tem mais de 2 banheiros?

In [200]:
living_2 = len (data.loc[((data['m2'] > 300) & (data['bathrooms'] > 2)), 'id'])
living_2

2201

In [201]:
print( 'A quantidade de casas que possuem mais de 300 metros quadrados de sala de estar e mais de 2 banheiros é de {}'.format( living_2 ))

A quantidade de casas que possuem mais de 300 metros quadrados de sala de estar e mais de 2 banheiros é de 2201
